Use `br` instead of a conditional when switching on a constant boolean

clubby789 opened the pull request:
r? @ghost

clubby789 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit b0be08c6a673577f8ae1a132528deb0ab3550cd8 with merge 14c7e36e619b79df407c2f6cf769648c6b7681cf...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b0be08c6a673577f8ae1a132528deb0ab3550cd8","merge_sha":"14c7e36e619b79df407c2f6cf769648c6b7681cf"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7776452346/job/21204621892)
Build commit: 14c7e36e619b79df407c2f6cf769648c6b7681cf (`14c7e36e619b79df407c2f6cf769648c6b7681cf`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7776452346/job/21204621892"},"merge_sha":"14c7e36e619b79df407c2f6cf769648c6b7681cf"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7776452346/job/21204621892)
Build commit: 14c7e36e619b79df407c2f6cf769648c6b7681cf (`14c7e36e619b79df407c2f6cf769648c6b7681cf`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7776452346/job/21204621892"},"merge_sha":"14c7e36e619b79df407c2f6cf769648c6b7681cf"} -->

rust-timer commented:
Queued 14c7e36e619b79df407c2f6cf769648c6b7681cf with parent 671eb38a4e5fadb75548841ec7ae31224eea0593, future [comparison URL](https://perf.rust-lang.org/compare.html?start=671eb38a4e5fadb75548841ec7ae31224eea0593&end=14c7e36e619b79df407c2f6cf769648c6b7681cf).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.1 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (14c7e36e619b79df407c2f6cf769648c6b7681cf): [comparison URL](https://perf.rust-lang.org/compare.html?start=671eb38a4e5fadb75548841ec7ae31224eea0593&end=14c7e36e619b79df407c2f6cf769648c6b7681cf&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=671eb38a4e5fadb75548841ec7ae31224eea0593&end=14c7e36e619b79df407c2f6cf769648c6b7681cf&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.8% | [-1.8%, -0.2%] | 17    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.8% | [-1.8%, -0.2%] | 17    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=671eb38a4e5fadb75548841ec7ae31224eea0593&end=14c7e36e619b79df407c2f6cf769648c6b7681cf&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.2% | [-3.7%, -0.8%] | 5     |
| Improvements ✅ <br /> (secondary) | -2.3% | [-2.5%, -2.1%] | 5     |
| All ❌✅ (primary)                 | -2.2% | [-3.7%, -0.8%] | 5     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=671eb38a4e5fadb75548841ec7ae31224eea0593&end=14c7e36e619b79df407c2f6cf769648c6b7681cf&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.9% | [-1.9%, -1.9%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.9% | [-1.9%, -1.9%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=671eb38a4e5fadb75548841ec7ae31224eea0593&end=14c7e36e619b79df407c2f6cf769648c6b7681cf&stat=size:linked_artifact)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.1% | [-2.6%, -0.1%] | 65    |
| Improvements ✅ <br /> (secondary) | -0.2% | [-0.3%, -0.0%] | 10    |
| All ❌✅ (primary)                 | -1.1% | [-2.6%, -0.1%] | 65    |
</details>

**Bootstrap**: 660.289s -> 662.794s (0.38%)
**Artifact size**: 308.12 MiB -> 308.09 MiB (-0.01%)

clubby789 commented:
Lets try applying this to all discriminants, not only bools
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

rust-timer commented:
Queued e99adce02fa787e0bd5af9c16309ef8fc5b6431e with parent 8c0b4f67c9b4bf477df38c16669fd576d46a2b3d, future [comparison URL](https://perf.rust-lang.org/compare.html?start=8c0b4f67c9b4bf477df38c16669fd576d46a2b3d&end=e99adce02fa787e0bd5af9c16309ef8fc5b6431e).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

clubby789 committed:
Use `br` instead of conditional when branching on constant

clubby789 commented:
r? compiler

fmease commented:
r? compiler

michaelwoerister commented:
Looks promising but I don't know enough about this to gauge if this has any subtle, undesirable side effects.

r? compiler

compiler-errors commented on compiler/rustc_codegen_ssa/src/mir/block.rs:
```diff
@@ -319,6 +319,14 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         let discr = self.codegen_operand(bx, discr);
+        let discr_value = discr.immediate();
+        // If our discriminant is a constant we can branch directly
+        if let Some(const_discr) = bx.const_to_opt_u128(discr_value, true) {
```
Why is `sign_ext` true?

clubby789 replied:
I wasn't quite sure what the right option there was. Should it be `false` or depend on the type of the discr?

compiler-errors replied:
Yeah, I think it depends on the type of discr? I really don't know though -- I recommend looking at how these switch targets are evaluated.

clubby789 replied:
I'll come back to it once perf has gone through for my two MIR-opt based approaches. They should allow us to waste less time emitting dead bbs to LLVM.

compiler-errors commented on compiler/rustc_codegen_ssa/src/mir/block.rs:
```diff
@@ -319,6 +319,14 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         targets: &SwitchTargets,
     ) {
         let discr = self.codegen_operand(bx, discr);
```
target_for_value is already being called below--the problem here is that it takes a u128, and consts are stored in the compiler as scalars in the MIR or machine-specific values here.

clubby789 replied:
Would checking if it's a constant then `Const::try_to_bits` be sufficient?

clubby789 replied:
Looking into it, it looks like at this point we've already codegenned the `i1 true`, as `if true` creates a local which has `true` assigned to it (which wouldn't be safe to remove without checking for other uses of the local).
`codegen_operand` here then doesn't actually talk to LLVM, we go down to `maybe_codegen_consume_direct` which loads the `i1 true` reference directly from `self.locals`

oli-obk replied:
> which loads the `i1 true` reference directly from `self.locals`

While that is true for `bool`, other types will need to generate an llvm value and load from it.

the8472 commented:
Perfect timing, In #120682 I have some additional code paths behind `if CONST { impl_a() } else { impl_b() }` and it's blowing up some compile benchmarks. If this can fix that I'll have to do less IR-shaving.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 978c0e01d5a95ead85a605844acd813c511d848b with merge 6a3fcc91d6c2793cdec9d313b8fd2bb8d6d48a3f...
<!-- homu: {"type":"TryBuildStarted","head_sha":"978c0e01d5a95ead85a605844acd813c511d848b","merge_sha":"6a3fcc91d6c2793cdec9d313b8fd2bb8d6d48a3f"} -->

rust-timer commented:
Queued 6a3fcc91d6c2793cdec9d313b8fd2bb8d6d48a3f with parent d6c46a23ce19e910225abacc33bcca9d0f549148, future [comparison URL](https://perf.rust-lang.org/compare.html?start=d6c46a23ce19e910225abacc33bcca9d0f549148&end=6a3fcc91d6c2793cdec9d313b8fd2bb8d6d48a3f).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (6a3fcc91d6c2793cdec9d313b8fd2bb8d6d48a3f): [comparison URL](https://perf.rust-lang.org/compare.html?start=d6c46a23ce19e910225abacc33bcca9d0f549148&end=6a3fcc91d6c2793cdec9d313b8fd2bb8d6d48a3f&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=d6c46a23ce19e910225abacc33bcca9d0f549148&end=6a3fcc91d6c2793cdec9d313b8fd2bb8d6d48a3f&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.2%  | [0.2%, 0.3%]   | 4     |
| Improvements ✅ <br /> (primary)   | -0.4% | [-0.4%, -0.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -0.2% | [-0.3%, -0.2%] | 2     |
| All ❌✅ (primary)                 | -0.4% | [-0.4%, -0.4%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=d6c46a23ce19e910225abacc33bcca9d0f549148&end=6a3fcc91d6c2793cdec9d313b8fd2bb8d6d48a3f&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.6%  | [0.1%, 7.5%]   | 4     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -5.4% | [-7.0%, -3.6%] | 3     |
| Improvements ✅ <br /> (secondary) | -2.2% | [-2.2%, -2.2%] | 1     |
| All ❌✅ (primary)                 | -0.3% | [-7.0%, 7.5%]  | 7     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=d6c46a23ce19e910225abacc33bcca9d0f549148&end=6a3fcc91d6c2793cdec9d313b8fd2bb8d6d48a3f&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.5% | [1.5%, 1.5%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.5% | [1.5%, 1.5%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=d6c46a23ce19e910225abacc33bcca9d0f549148&end=6a3fcc91d6c2793cdec9d313b8fd2bb8d6d48a3f&stat=size:linked_artifact)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1%  | [0.0%, 0.2%]   | 6     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.3% | [-0.6%, -0.1%] | 32    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.2% | [-0.6%, 0.2%]  | 38    |
</details>

**Bootstrap**: 662.763s -> 662.191s (-0.09%)
**Artifact size**: 308.24 MiB -> 308.22 MiB (-0.01%)

clubby789 commented:
Reverted to the original version so I'm removing the regression label

the8472 commented:
To check my understanding for the purpose of optimizing library code:

Given

```rust
fn foo<T: SomeTrait, const C: bool>(x: T)
{
  if C {
    <T as SomeTrait>::a(x);
  } else {
    <T as SomeTrait>::b(x);
  }
}
```

Which work on the non-taken branch _and its call-tree_ would be avoided? Emitting llvmir, calculating generic `const {}` blocks, some other stuff?

the8472 commented:
Ah, that's less than I was looking for. The CopyProp one also wouldn't help in cases where the const only becomes known at monomorphization time, right?

saethlin commented:
r? saethlin

I'll review this by the end of the week.

saethlin commented:
As opposed to all the other kinds of tweaks I want to make around lowering `SwitchInt(const _)` which may have knock-on effects such as not reporting post-mono errors, I'm pretty confident this change is perf-only.

@bors r+

bors commented:
:pushpin: Commit 86ad91f03bdc490b2bb569b50866c6839e70bd57 has been approved by `saethlin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=saethlin 86ad91f03bdc490b2bb569b50866c6839e70bd57 -->
<!-- homu: {"type":"Approved","sha":"86ad91f03bdc490b2bb569b50866c6839e70bd57","approver":"saethlin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 86ad91f03bdc490b2bb569b50866c6839e70bd57 with merge e8bd826cd4c025b51ef7fb18eb31d613ffab5812...
<!-- homu: {"type":"BuildStarted","head_sha":"86ad91f03bdc490b2bb569b50866c6839e70bd57","merge_sha":"e8bd826cd4c025b51ef7fb18eb31d613ffab5812"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/8011975917/job/21886335507) [(plain)](https://github.com/rust-lang-ci/rust/commit/e8bd826cd4c025b51ef7fb18eb31d613ffab5812/checks/21886335507/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[endgroup]
[TIMING] core::build_steps::test::Compiletest { compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu }, target: x86_64-unknown-linux-gnu, mode: "mir-opt", suite: "mir-opt", path: "tests/mir-opt", compare_mode: None } -- 31.535
[TIMING] core::build_steps::test::MirOpt { compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu }, target: x86_64-unknown-linux-gnu } -- 0.025
##[group]Testing stage1 compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu)
error: detected legacy-style directives in compiletest test: /checkout/tests/codegen/constant-branch.rs:1, please use `ui_test`-style directives `//@` instead:Some(
        None,
        None,
        "compile-flags: -Zmir-opt-level=0 -C no-prepopulate-passes -Copt-level=0",
)
errors encountered during EarlyProps parsing: /checkout/tests/codegen/constant-branch.rs
thread 'main' panicked at src/tools/compiletest/src/header.rs:74:13:
errors encountered during EarlyProps parsing
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8011975917/job/21886934046)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/8011975917/job/21886934046","builder_name":"checks-actions"} -->

bors commented:
:pushpin: Commit 15dad2b36f21c0d0e876510e3abb1beec719905c has been approved by `saethlin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=saethlin 15dad2b36f21c0d0e876510e3abb1beec719905c -->
<!-- homu: {"type":"Approved","sha":"15dad2b36f21c0d0e876510e3abb1beec719905c","approver":"saethlin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 15dad2b36f21c0d0e876510e3abb1beec719905c with merge 77899b30fce14e9fa4565f5a963161d19ee5cf1e...
<!-- homu: {"type":"BuildStarted","head_sha":"15dad2b36f21c0d0e876510e3abb1beec719905c","merge_sha":"77899b30fce14e9fa4565f5a963161d19ee5cf1e"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-nopt`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/8012632507/job/21888292760) [(plain)](https://github.com/rust-lang-ci/rust/commit/77899b30fce14e9fa4565f5a963161d19ee5cf1e/checks/21888292760/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8012632507/job/21889353449)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/8012632507/job/21889353449","builder_name":"checks-actions"} -->

clubby789 commented:
Looks like `sign_ext` indeed needs to be `false` - discriminants are stored sign-extended to their type size (https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/util/struct.Discr.html#structfield.val). Updated and added a new codegen test but this should probably be re-checked to make sure I'm right on this
@rustbot ready

saethlin commented:
o7
@bors r+

bors commented:
:pushpin: Commit 7159aed51e7ef4d8459a10e449bf8e9cc09c9b98 has been approved by `saethlin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=saethlin 7159aed51e7ef4d8459a10e449bf8e9cc09c9b98 -->
<!-- homu: {"type":"Approved","sha":"7159aed51e7ef4d8459a10e449bf8e9cc09c9b98","approver":"saethlin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:evergreen_tree: The tree is currently [closed](https://github.com/rust-lang/rust/pull/121523#issuecomment-1962124798) for pull requests below priority 50. This pull request will be tested once the tree is reopened.

bors commented:
:hourglass: Testing commit 7159aed51e7ef4d8459a10e449bf8e9cc09c9b98 with merge 89d8e3116ccb7f55760825b731de4f4e0fd76770...
<!-- homu: {"type":"BuildStarted","head_sha":"7159aed51e7ef4d8459a10e449bf8e9cc09c9b98","merge_sha":"89d8e3116ccb7f55760825b731de4f4e0fd76770"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8034607678/job/21947624878)
Approved by: saethlin
Pushing 89d8e3116ccb7f55760825b731de4f4e0fd76770 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"saethlin","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8034607678/job/21947624878"},"merge_sha":"89d8e3116ccb7f55760825b731de4f4e0fd76770"} -->

rust-timer commented:
Finished benchmarking commit (89d8e3116ccb7f55760825b731de4f4e0fd76770): [comparison URL](https://perf.rust-lang.org/compare.html?start=2ae1bb671183a072b54ed8ed39abfcd72990a3e7&end=89d8e3116ccb7f55760825b731de4f4e0fd76770&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=2ae1bb671183a072b54ed8ed39abfcd72990a3e7&end=89d8e3116ccb7f55760825b731de4f4e0fd76770&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.9% | [-2.2%, -0.1%] | 19    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.9% | [-2.2%, -0.1%] | 19    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=2ae1bb671183a072b54ed8ed39abfcd72990a3e7&end=89d8e3116ccb7f55760825b731de4f4e0fd76770&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -3.6% | [-6.1%, -1.1%] | 7     |
| Improvements ✅ <br /> (secondary) | -4.0% | [-4.0%, -4.0%] | 1     |
| All ❌✅ (primary)                 | -3.6% | [-6.1%, -1.1%] | 7     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=2ae1bb671183a072b54ed8ed39abfcd72990a3e7&end=89d8e3116ccb7f55760825b731de4f4e0fd76770&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.0% | [-2.0%, -2.0%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=2ae1bb671183a072b54ed8ed39abfcd72990a3e7&end=89d8e3116ccb7f55760825b731de4f4e0fd76770&stat=size:linked_artifact)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.7% | [-3.8%, -0.0%] | 69    |
| Improvements ✅ <br /> (secondary) | -0.6% | [-1.1%, -0.3%] | 21    |
| All ❌✅ (primary)                 | -1.7% | [-3.8%, -0.0%] | 69    |
</details>

**Bootstrap**: 650.564s -> 651.95s (0.21%)
**Artifact size**: 311.05 MiB -> 311.05 MiB (-0.00%)
